### PR TITLE
Allow creating `ComplexConstant_` in `DCMPLX` for compile time arguments

### LINKED ISSUE #1432: SciPy: "ComplexConstructor with runtime arguments not implemented yet"
While working on #1416, I came across this issue that produces the same error but is actually different ( compile time vs runtime ).
```fortran
SUBROUTINE CPDSA()
    print *, DCMPLX(1.0D0, 6.0D0)
ENDSUBROUTINE

program main
    call CPDSA()
end program
```

> LFortran 

```console
$ lfortran ./a.f90 --implicit-typing
code generation error: ComplexConstructor with runtime arguments not implemented yet.


Note: if any of the above error or warning messages are not clear or are lacking
context please report it to us (we consider that a bug that needs to be fixed).
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #1432.

### CODE REVIEW

File: tests/reference/asr-complex_07-8d914e3.stdout [hunk 1]
```diff
@@ -250,7 +250,11 @@
                                 (Real 8 [])
                             )
                             (Complex 8 [])
-                            ()
+                            (ComplexConstant
+                                3.140000
+                                0.000000
+                                (Complex 8 [])
```
Comment (Pranavchiku) [inline tests/reference/asr-complex_07-8d914e3.stdout:256]: This change looks good to me, I cross-verified it.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 2]
```diff
@@ -3107,7 +3107,20 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
         }
         ASR::ttype_t *type = ASRUtils::TYPE(ASR::make_Complex_t(al, x.base.base.loc,
                                 kind_value, nullptr, 0));
-        return ASR::make_ComplexConstructor_t(al, x.base.base.loc, x_, y_, type, nullptr);
+        ASR::expr_t* x_value = ASRUtils::expr_value(x_);
```
[the remark this answers is not in the document]
Comment (czgdp1807) [reply src/lfortran/semantics/ast_common_visitor.h:3110]: Why do you want to create a `ComplexConstant` from a variable complex number? In, `DCMPLX(PD, 0.0D0)`, `PD` is a variable, so we should create a `ComplexConstructor` node? Why `ComplexConstant` node creation is necessary here?

[the remark this answers is not in the document]
Comment (Pranavchiku) [reply src/lfortran/semantics/ast_common_visitor.h:3110]: I am trying to create a `ComplexConstant` expression which can be set as `m_value` of `ComplexConstructor` node.

[the remark this answers is not in the document]
Comment (czgdp1807) [reply src/lfortran/semantics/ast_common_visitor.h:3110]: If `ComplexConstant` can not be created that means compile time value of `ComplexConstructor` can not reduce to compile time constant. So you can set `ComplexConstructor::m_value` to `nullptr`.

[the remark this answers is not in the document]
Comment (czgdp1807) [reply src/lfortran/semantics/ast_common_visitor.h:3110]: I think `m_value` is an optional attribute so should be used only when its not `nullptr`.

**Comment by czgdp1807** [comment]
Is it ready for review?

**Comment by Pranavchiku** [comment]
Yeah

File: src/lfortran/semantics/ast_common_visitor.h [hunk 3]
```diff
@@ -3107,7 +3107,20 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+        ASR::expr_t* cc_expr = nullptr;
+        double x_value_ = 0.0;
+        double y_value_ = 0.0;
+        if (x_value && y_value && ASR::is_a<ASR::RealConstant_t>(*x_value) && ASR::is_a<ASR::RealConstant_t>(*y_value)) {
+            ASR::RealConstant_t* x_real = ASR::down_cast<ASR::RealConstant_t>(x_value);
+            x_value_ = x_real->m_r;
+            ASR::RealConstant_t* y_real = ASR::down_cast<ASR::RealConstant_t>(y_value);
+            y_value_ = y_real->m_r;
+            cc_expr = ASRUtils::EXPR(ASR::make_ComplexConstant_t(al, x.base.base.loc,
+                                                                 x_value_, y_value_, type));
```
Comment (czgdp1807) [inline src/lfortran/semantics/ast_common_visitor.h:3121]: Use, for extracting real values. It returns `false` if value cannot be extracted in which case you can set `ComplexConstructor::m_value` to `nullptr`.

https://github.com/lfortran/lfortran/blob/f5ccf212d7d19a5193856770a7fed846284eacec/src/libasr/asr_utils.h#L805

**Review by czgdp1807 — CHANGES_REQUESTED** [review]

File: integration_tests/complex_11.f90 [hunk 4]
```diff
@@ -0,0 +1,7 @@
+SUBROUTINE CPDSA()
+    print *, DCMPLX(1.0D0, 6.0D0)
```
Comment (czgdp1807) [inline integration_tests/complex_11.f90:2]: Would be good to do `if () error stop` checks here.

**Review by czgdp1807 — CHANGES_REQUESTED** [review]

**Review by czgdp1807 — APPROVED** [review]

**Pull request merged**